sips/sip-015/sip-015-network-upgrade.md

```diff
@@ -0,0 +1,1644 @@
+can use `stack-aggregation-commit` to commit to each PoX reward
+address separately.
+
+## Clarity
+
+If this SIP is ratified, then at the time of the
+Stacks 2.1 network upgrade, the Clarity smart
+contract language will be expanded with new features and some new
+behaviors for existing native methods. In order to support these
+changes, this SIP proposes to introduce *Clarity versioning for smart
+contracts*. In this scheme, each smart contract will be associated
+with a particular Clarity version. The execution environment will
+track the current version for a given execution (in the
+implementation, this will be via the `ContractContext`), and use that
+to select which features are available, and which native method
+implementations will be used. Clarity 2 contracts can invoke Clarity 1
+contracts (and vice-versa), but particular care will need to be taken
+if a new native keyword is used in the Clarity 1 contract's API.
+
+For example:
+
+```
+Contract A (Clarity 1 Contract):
+  (define-public (stx-account) ...)
+  
+Contract B (Clarity 2 Contract):
+  (contract-call contract-A stx-account)
+```
+
+In such cases, the new keyword (in the above example, the keyword is
+`stx-account`) *cannot* be used by the Clarity 2 contract, and
+therefore invoking the public method of Contract A from Contract B is
+not allowed. To address these cases, contract authors must launch a
+Clarity 1 contract that interposes on Contract A, providing a Clarity
+2 compatible interface.
+
+Existing, pre-2.1 contracts will all be Clarity 1. New contracts will
+default to Clarity 2, and a new Stacks transaction wire format for publishing
+smart contracts will allow contract publishers to choose the version that
+their contract should use.
+
+### New method: `stx-account`
+
+* **Input Signature:** `(stx-account (principal-address principal))`
+* **Output Signature:** `{ locked: uint, unlock-height: uint, unlocked: uint }`
+
+This method returns the status of STX account: the account's
+current STX balance, the amount of STX that is currently locked,
+and the unlock height for the account.
+
+**Rationale:**  This method will be used by the `pox-2` contract to validate
+various contract-calls: implementing early unlocks, lock extensions,
+lock increases, etc.  It exposes PoX lock state to smart contracts, which is not
+entirely visible even in `pox` or `pox-2`.
+
+**Examples:**
+
+```clarity
+(stx-account 'SZ2J6ZY48GV1EZ5V2V5RB9MP66SW86PYKKQ9H6DPR) ;; Returns (tuple (locked u0) (unlock-height u0) (unlocked u0))
+(stx-account (as-contract tx-sender)) ;; Returns (tuple (locked u0) (unlock-height u0) (unlocked u1000))
+```
+
+### New method: `principal-destruct`
+
+* **Input Signature:** `(principal-destruct (principal-address principal))`
+* **Output Signature:** `(response { hash-bytes: (buff 20), name: (optional (string-ascii 40)), version: (buff 1) } { hash-bytes: (buff 20), name: (optional (string-ascii 40)), version: (buff 1) })`
```

Comment:
According to description this function serves two purposes:
1) decompose principal to its parts
2) validate if its version byte matches the network

Every function should have only one responsibility. Therefore it would make more sense to split this one it into two separate functions:

#### principal-destruct
* **Input Signature**: `(principal-destruct (principal-address principal))`
* **Output Signature**: `{ hash-bytes: (buff 20), name: (optional (string-ascii 40)), version: (buff 1) }`

#### is-network-principal
* **Input Signature**: `(principal-destruct (principal-address principal))`
* **Output Signature**: `bool`

Reply:
First, `is-standard` validates the version byte independently.

Second, the reason `principal-destruct` and `principal-construct` return an `(err ..)` if the version byte is incompatible with the network is to set developers up for success, and in doing so, reduce the chance of harm being brought to users.  If the history of smart contract bugs is any indication, developers can't be relied upon to do a good job of validating untrusted input (in which case, _users_ stand to _suffer tremendously_ from the resulting loss or theft of funds).  By making these two functions perform validation automatically, we make it so that the developer _cannot avoid_ validating the input.

Reply:
`is-standard` does more than that :wink: https://github.com/stacksgov/sips/pull/95#discussion_r1007254146

If you want to keep developers safe, then new data type (ie. `generic-principal` or `stx-principal` or `valid-principal`) that represents principal which is valid and belongs to any network should be introduced. New data type plus extra conversion function (ie. `valit-principal-to-principal`) that returns `(optional principal)` or `(response principal uint)`. That way you'll enforce proper rules on a language level and make impossible (or almost impossible) to deploy a contract that break these rules.

By introducing new data-type developers won't be able to shoot themselves in the foot because you'll take the gun out of their hands. By wrapping something with `(err ..)` you just telling them "Hey! Be careful. You might hurt yourself with that"


On a side note, I though that it is impossible to pass principal from different network and for example send STX from mainnet to testnet address, or store testnet address in mainnet contract variable. I thought such transaction would be considered as invalid and fail or be un-mineable.

Reply:
The goals are to keep users safe and developers happy.  If they are in conflict, the latter can be sacrificed to preserve the former.

Introducing a bunch of new types doesn't keep users any safer than this function's current behavior, but it would make developers miserable.  Not only would they have to reason about more types, but also it would _de facto_ prevent Clarity 2 contracts from calling into Clarity 1 contracts and vice versa, because Clarity 1 doesn't know about these new types.  So as far as I'm concerned, adding new types to deal with this problem is a non-starter.

> On a side note, I though that it is impossible to pass principal from different network and for example send STX from mainnet to testnet address, or store testnet address in mainnet contract variable. I thought such transaction would be considered as invalid and fail or be un-mineable.

You cannot instantiate a smart contract with a non-standard address, but you can send to a non-standard address.  Making this function return an error when the address is non-standard preserves compatibility with the current system while also de-risking the chance that developers permit it by mistake.

Reply:
> Introducing a bunch of new types doesn't keep users any safer than this function's current behavior, but it would make developers miserable. Not only would they have to reason about more types, but also it would de facto prevent Clarity 2 contracts from calling into Clarity 1 contracts and vice versa, because Clarity 1 doesn't know about these new types. So as far as I'm concerned, adding new types to deal with this problem is a non-starter.

I strongly disagree with all your arguments.
Why it would make developers miserable? Because they need to familiarize themselves with one extra type that would make their code more safe? Why would it prevent from calling V2 into V1 and vice versa?

**Small example:**
Let's assume for a moment that you introduced new type `stx-principal` and function `to-principal` with following signature:
```clarity
(to-principal stx-principal (response principal uint))
```
, and that `principal-construct` returns `(response stx-principal uint)`.

Can we write a code that will be full secured by language? I think we can:
```clarity
(define-public (test (version (buff 1)) (pub-key-hash (buff 20)) (contract-name (optional (string-ascii 40)))) 
  (match (principal-construct version pub-key-hash contract_name) 
    ;; we were able to construct valid principal so we "store it" in local variable of stx-principal type
    target-stx-principal 
    (match (to-principal target-stx-principal) 
        ;; we managed to successfully convert stx-principal to principal that is valid within our network
        ;; let's store it in another local variable and then use it
        target-principal
        ;; so we can execute transfer knowing that `target-principal` is an address that matches our network
        (stx-transfer? u100 tx-sender target-principal)
        ;; unfortunately we were not able to convert stx-principal to principal so we store error in variable
        err-value
        ;; and return our own custom error
        (err "invalid principal")
    )
    ;; failed to construct principal
    (err "invalid data")
  ) 
)
```
Can we utilize `stx-principal` and somehow call functions in V1 contract? I think so: 
```clarity
(define-public (test (version (buff 1)) (pub-key-hash (buff 20)) (contract-name (optional (string-ascii 40)))) 
  (match (principal-construct version pub-key-hash contract_name) 
    ;; we were able to construct valid principal so we "store it" in local variable of stx-principal type
    target-stx-principal 
    (match (to-principal target-stx-principal) 
        ;; we managed to successfully convert stx-principal to principal that is valid withing our network
        target-principal
        ;; and now we call call V1 contract knowing that principal we are passing is 100% valid
        (contract-call? .clarity-v1-contract my-function target-principal)
        ;; unfortunately we were not able to convert stx-principal to principal so we store error in variable
        err-value
        ;; and return our own custom error
        (err "invalid principal")
    )
    ;; failed to construct principal
    (err "invalid data")
  ) 
)
```
Can we call a function defined in V2 contract from V1 contract? Why would anyone think that it would be impossible? Both versions recognize `principal` type, so it will work the same way as it is working right now.
```clarity
;; v1-contract
(define-trait bla-trait
  ((call-me (principal uint) (response bool uint)))
)

(define-public (test (tgt <bla-trait>) (who principal) (x uint))
    (contract-call? txt call-me who x)
)

;; v2-contract
(define-public (call-me (who principal) (x uint))
  (begin
    (asserts! (> x u0) (err u123))
    (print who)
    (print x)
    (ok true)
  )
)

;; execution:
(contract-call? .v1-contract test .v2-contract tx-sender u200)
```
What if you want to use `principal-destruct` in V2 contract but you want to call it from V1? According to my suggestion `principal-destruct` should accept only `stx-principal` type argument, right? And V1 doesn't know anything about `stx-principal`. 
What can we do?
The same thing we do with almost any conversion. If we can convert X to Y, then we should be able (with some exceptions) to convert also Y to X.  You can introduce `to-stx-principal` function with following signature and problem will be solved:
```clarity
(to-stx-principal principal principal)
```
Add something like `(is-network-principal principal bool)` that is able to tell if principal provided is valid withing your network. And enforce that in V2 all functions like `stx-transfer/burn?`, `ft-transfer/mint/burn`, `nft-transfer/mint/burn` when called will accept only principals valid withing your network.

I believe that if you do something similar to what is described above you will give yourself a chance to make it **virtually impossible** to create a contract that will allow users to send funds to address that belongs to a different network. 
And I think it would make both developers and users happy.

Reply:
As I said earlier, one of the goals of `principal-construct` and `principal-destruct` is to make it impossible for developers to forget to validate the input to these functions.  This is achieved representing valid input as an `(ok ..)` and invalid input as an `(err ..)`.

In your proposal, developers would always need to remember to (1) check that a principal is valid on the network after calling `principal-construct`, and (2) remember to convert `stx-principal` to and from Clarity 1's `principal` type when calling into Clarity 1.  Also, since it seems `stx-principal` and `principal` are distinct types in Clarity 2 that can receive assets (I'm not sure what the real difference is?), then the developer would additionally be responsible for (3) determining which principal sub-type a given principal value is in order to decide whether or not to send assets to it.

That's three separate things that this proposal now compels developers to always remember to do manually.

> I believe that if you do something similar to what is described above you will give yourself a chance to make it virtually impossible to create a contract that will allow users to send funds to address that belongs to a different network.
And I think it would make both developers and users happy.

This can be addressed separately: simply require all the `*-transfer?` and `*-mint?` built-ins to fail if the address does not have a network-standard version byte.

This was considered for inclusion in Stacks 2.1 but was ultimately slated for a future hard fork (https://github.com/stacks-network/stacks-blockchain/issues/3118), since it was unclear that this would actually add anything.  It's already the case in 2.1 that all of the ways to create a `principal` type from non-principal data will fail if the input data would not lead to the construction of a network-compatible `principal` value (this applies to `principal-construct`, `principal-destruct`, and `principal-of`).  Also, AFAIK all of the Stacks wallets and exchanges prevent you from using a non-standard principal.  You could write a smart contract that uses a non-standard principal or manually craft a transaction that called a function with a non-standard principal value, but in both cases you'd be doing so deliberately.

Reply:
> In your proposal, developers would always need to remember to (1) check that a principal is valid on the network after calling principal-construct, and (2) remember to convert stx-principal to and from Clarity 1's principal type when calling into Clarity 1.

In reality tools will inform developers that they are trying to pass value with incorrect type. The same way they are informing us right now when we try to pass `trait-ref` instead of `principal`, `uint` instead of `int`, raw value instead of response etc. Type system will take care of that.

> Also, since it seems `stx-principal` and `principal` are distinct types in Clarity 2 that can receive assets (I'm not sure what the real difference is?)

No, those are two different types just like `principal` and `trait-ref`. `stx-principal` represents any syntax wise valid principal (regardless of which network it is associated with) and it can't receive assets (never). While `principal` is any syntax wise valid principal with byte version that mach the network on which contract is executed, and therefore it can safely receive and spend assets. 

In other words you can say that `stx-principal` is like a string with build it validation that verifies if it mach certain pattern or not. And `principal` is a class/object/struct that holds internally `stx-principal`, it is able to tell if it is valid within a particular network, and can send and receive assets.

Reply:
> In reality tools will inform developers that they are trying to pass value with incorrect type. The same way they are informing us right now when we try to pass trait-ref instead of principal, uint instead of int, raw value instead of response etc. Type system will take care of that.

Since `principal-construct` and `principal-destruct` return a `response`, instead of a `principal`, developers can be assured that they will be informed by the tooling if they do not unwrap the `response` to get at the inner data.  That's the reason this API is written the way it is.  So, it seems our two approaches achieve the same end.  However, I continue to believe that mine is superior because it does not introduce any new types or extensions to the type system to achieve its goal.

Reply:
You want this and few other functions to work with all instances of Stacks network, and I get that, but your approach will force users to write very weird code.

For me your reasoning is convoluted and it will make language convoluted and ambiguous.
Instead of having clearly defined functions that does one thing and one thing only, we'll have functions that does 2 things and have 3 different results:
- success
- semi error (we managed to de-structure address but it is from a different network)
- error (address can't be de-structured)

Reasoning you have for this and `principal-construct` is not aligned with other functions. It is not even aligned with new `is-standard` function. 
Here is why: If I'll want to find out if principal from any network is standard or not I can't use `is-standard` to do that. I have to use `principal-destruct`.  

> So, it seems our two approaches achieve the same end.

I want the problem to be eliminated on the language level. Make it impossible to send funds to an address from a different network. You want to make it preventable and still fully rely on developers and users.
One approach is a permanent, probably a foolproof fix, another one is just a band-aid.  

> Generally speaking, blockchain code is append-only. There is no way to delete previous Clarity functionality, because the implementation requires it to validate previous blocks. Therefore, we should always prioritize API designs that minimize the amount of new functionality added to the VM, since it must be maintained forever.

I'm not asking you to delete any previous Clarity functionalities.
In V1 `principal-of?` has a bug that will be fixed from 2.1 onward. You're not going to delete it and implement something brand new. And I believe you can do exactly the same with how `principal` is handled.

Reply:
> Make it impossible to send funds to an address from a different network.

If this is your goal all along, then this would be better achieved by making it so all of the `-mint?` and `-transfer?` functions simply returned an error when given an unsupported address.  This wouldn't require introducing any new types either.  This wasn't proposed in this particular SIP, but I think the community would be receptive to such a proposal.  There's even an issue already tracking it: https://github.com/stacks-network/stacks-blockchain/issues/3118.

> I'm not asking you to delete any previous Clarity functionalities.

I realize that.  I'm instead arguing that a simpler solution that does not add new types creates a lower maintenance burden.  Adding new types would not only require more code, but also make calls between Clarity 1 and Clarity 2 much more difficult to do correctly.

Reply:
Until a few days ago I wasn't aware that Clarity have such massive flaw and it allows people to send funds to addresses that belongs to a different network. And it makes me wonder how on earth it has not been communicated anywhere? Why there is no single word about it in documentation? In short WTF? 

I suggested new type when I've found out why you decided to gave this function two distinctive responsibilities, and to show you that there is a much cleaner solution that will provide you the same if not greater safety. A solution that does not require returning semi-errors. 

What is my primary goal? 
1. Fix all `*-mint?` and `*-transfer?` functions. If it is not doable right now - add proper warning to documentation using big, bold red font and then fix it ASAP.
2. Remove version byte check from `principal-destruct`, `principal-construct`, `is-standard` and make all 3 of them do **exactly** what their names imply and nothing else.
3. Add new function that validates version byte

That way If developers want destruct or construct principal, or check if it is standard one or not, they use functions that does what they need. If they want to verify if `principal` belongs to the network - they use function introduced in 3rd point. 
Everything is fully composable. There is no need to handle errors differently based on whether we are or aren't interested in network check. If we want to check if principal is standard or not regardless of the network it belongs to, then there is no need to use `principal-destruct` - you simply use `is-standard` - plain and simple.

If you document these functions (`principal-destruct`, `principal-construct`, `is-standard` and new one from 3rd point) properly, then developers will quickly learn how to use them and how to do it securely (nobody needs another: https://github.com/stacks-network/docs/issues/1249#issuecomment-885185172).